{
  "title": "Code, people and fallibility",
  "date": "2022-04-22T00:00:00.000Z",
  "lede": "Tech failures, like Facebook's DNS mishap and Atlassian's outage, highlight a key insight — it's not just about the code, but the humans behind it. These incidents underscore the importance of communication and collaboration in technology, reminding us that at the core of every technical glitch is a human error waiting to be understood and addressed.",
  "category": "writing",
  "tags": [
    "teamwork",
    "engineering"
  ],
  "libraryType": "post",
  "body": {
    "raw": "\nLast year, Facebook accidentally overwrote its own DNS records leading to a “cascade of errors” and worldwide outage for all its platforms. The outage was likely caused by Facebook network engineers accidentally locking themselves out of the larger configuration system during an update. [This quote](https://news.ycombinator.com/item?id=28749244), purportedly from a member of the recovery team, is particularly insightful:\n\n> …the people with physical access is separate from the people with knowledge of how to actually authenticate to the systems and people who know what to actually do, so there is now a logistical challenge with getting all that knowledge unified.\n\nIn the last fortnight, [Atlassian had the longest outage they’ve ever had](https://newsletter.pragmaticengineer.com/p/scoop-atlassian?s=r). What started out as “downtime” on April 4, turned into 10 days of complete shutdown. And they took until [day 9 to formally communicate with their customers](https://twitter.com/SriViswan/status/1514072080181956608?s=20&t=VcElOD_KVnRCdgNOLeB5Ig). We can only guess at the chaos behind the scenes.\n\nOf course, we can say that’s Big Tech, they should have the resources and bright people to figure this out and stop this happening. Shame on them. Or we can say that that's Big Tech, that's different to the scale of what we do. It would never happen to us. And yet if we pause for a minute and consider all the mundane moments that occurred in the prior days, months and perhaps years that lead to these incidents, we may gain a unitary insight into _how organisations of people work with code_.\n\n## The world of code\n\nA piece of code takes an input to produce an output and if you add the wrong input, that code itself is none the wiser. But just above that piece of code, we have a world in which the human organisation and the people involved provide the wherewithal for the code to be written into a codebase: what it’s for, how it works, its dependent inputs, the chain of its related outputs and so on.\n\nNow picture in your mind how that codebase relates to the larger system at play: the intent of the people in the organisation, their multiple, mixed and shared drivers for what they do, how they co-operate and the end customers they’re beholden to.\n\nIf we consider the world of code in which the Atlassian Incident took place, we might ask ourselves a few simple questions: did they safeguard the kind of input for this piece of code? Were the people doing the work tired and simply made a mistake? Were there pre-existing issues that had been acknowledged among the group? Was this communicated and with which group?\n\nOnce we consider questions like these, it is easy to see that _the code itself is but one small piece in a larger puzzle that is mostly about human agency and co-operation_.\n\nBecause it is so malleable, we can do all kinds of new things with code, some of which have been [surprisingly valuable](http://web.archive.org/web/19980610100009/www.redherring.com/mag/issue55/economics.html) for the communications and business spheres in the past three decades. And precisely because it is so malleable, code always contains the possibility of surprising us towards the other side of the reward spectrum: _risk_.\n\n## The human code\n\nBeyond a certain point—as Atlassian and Facebook will attest to—what is complicated can turn into unbounded complexity. It is easy to gawk at the runaway train of incidents resulting from such complexity until you too wish to make use of what code might do for you. Then you will engage in your own “world of code”: a place where code does exactly what you and your organisation tell it to, in order to reap the returns across the risk/reward spectrum.\n\nHow then do we mitigate risk in the world of code? Our best advice is to understand that _behind the code is always a fallible human who understands how the code works_. And that human has to co-ordinate and communicate with other fallible humans around their ideas of what that code should do.\n\nIn such scenarios, we should be humble not just about our ability to programme complicated systems, but also about our ability to mutually understand, trust and co-ordinate with each-other.\n\nA unitary insight into working with code is that it’s not about code, it’s about people.\n\n<Note>\n  This article was originally published on the [Grade\n  Substack](https://grade.substack.com/p/code-people-and-fallibility).\n</Note>\n",
    "code": "var Component=(()=>{var lt=Object.create;var D=Object.defineProperty;var ut=Object.getOwnPropertyDescriptor;var dt=Object.getOwnPropertyNames;var ct=Object.getPrototypeOf,ft=Object.prototype.hasOwnProperty;var z=(u,n)=>()=>(n||u((n={exports:{}}).exports,n),n.exports),pt=(u,n)=>{for(var p in n)D(u,p,{get:n[p],enumerable:!0})},ge=(u,n,p,y)=>{if(n&&typeof n==\"object\"||typeof n==\"function\")for(let v of dt(n))!ft.call(u,v)&&v!==p&&D(u,v,{get:()=>n[v],enumerable:!(y=ut(n,v))||y.enumerable});return u};var mt=(u,n,p)=>(p=u!=null?lt(ct(u)):{},ge(n||!u||!u.__esModule?D(p,\"default\",{value:u,enumerable:!0}):p,u)),ht=u=>ge(D({},\"__esModule\",{value:!0}),u);var xe=z((Rt,Re)=>{Re.exports=React});var ke=z(G=>{\"use strict\";(function(){\"use strict\";var u=xe(),n=Symbol.for(\"react.element\"),p=Symbol.for(\"react.portal\"),y=Symbol.for(\"react.fragment\"),v=Symbol.for(\"react.strict_mode\"),K=Symbol.for(\"react.profiler\"),X=Symbol.for(\"react.provider\"),H=Symbol.for(\"react.context\"),T=Symbol.for(\"react.forward_ref\"),A=Symbol.for(\"react.suspense\"),F=Symbol.for(\"react.suspense_list\"),O=Symbol.for(\"react.memo\"),I=Symbol.for(\"react.lazy\"),Oe=Symbol.for(\"react.offscreen\"),J=Symbol.iterator,Ce=\"@@iterator\";function Se(e){if(e===null||typeof e!=\"object\")return null;var t=J&&e[J]||e[Ce];return typeof t==\"function\"?t:null}var R=u.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;function m(e){{for(var t=arguments.length,r=new Array(t>1?t-1:0),a=1;a<t;a++)r[a-1]=arguments[a];Pe(\"error\",e,r)}}function Pe(e,t,r){{var a=R.ReactDebugCurrentFrame,l=a.getStackAddendum();l!==\"\"&&(t+=\"%s\",r=r.concat([l]));var d=r.map(function(i){return String(i)});d.unshift(\"Warning: \"+t),Function.prototype.apply.call(console[e],console,d)}}var je=!1,Ue=!1,De=!1,Ae=!1,Fe=!1,Z;Z=Symbol.for(\"react.module.reference\");function Ie(e){return!!(typeof e==\"string\"||typeof e==\"function\"||e===y||e===K||Fe||e===v||e===A||e===F||Ae||e===Oe||je||Ue||De||typeof e==\"object\"&&e!==null&&(e.$$typeof===I||e.$$typeof===O||e.$$typeof===X||e.$$typeof===H||e.$$typeof===T||e.$$typeof===Z||e.getModuleId!==void 0))}function We(e,t,r){var a=e.displayName;if(a)return a;var l=t.displayName||t.name||\"\";return l!==\"\"?r+\"(\"+l+\")\":r}function Q(e){return e.displayName||\"Context\"}function w(e){if(e==null)return null;if(typeof e.tag==\"number\"&&m(\"Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue.\"),typeof e==\"function\")return e.displayName||e.name||null;if(typeof e==\"string\")return e;switch(e){case y:return\"Fragment\";case p:return\"Portal\";case K:return\"Profiler\";case v:return\"StrictMode\";case A:return\"Suspense\";case F:return\"SuspenseList\"}if(typeof e==\"object\")switch(e.$$typeof){case H:var t=e;return Q(t)+\".Consumer\";case X:var r=e;return Q(r._context)+\".Provider\";case T:return We(e,e.render,\"ForwardRef\");case O:var a=e.displayName||null;return a!==null?a:w(e.type)||\"Memo\";case I:{var l=e,d=l._payload,i=l._init;try{return w(i(d))}catch{return null}}}return null}var g=Object.assign,E=0,ee,te,re,ne,ae,oe,ie;function se(){}se.__reactDisabledLog=!0;function Ve(){{if(E===0){ee=console.log,te=console.info,re=console.warn,ne=console.error,ae=console.group,oe=console.groupCollapsed,ie=console.groupEnd;var e={configurable:!0,enumerable:!0,value:se,writable:!0};Object.defineProperties(console,{info:e,log:e,warn:e,error:e,group:e,groupCollapsed:e,groupEnd:e})}E++}}function Ye(){{if(E--,E===0){var e={configurable:!0,enumerable:!0,writable:!0};Object.defineProperties(console,{log:g({},e,{value:ee}),info:g({},e,{value:te}),warn:g({},e,{value:re}),error:g({},e,{value:ne}),group:g({},e,{value:ae}),groupCollapsed:g({},e,{value:oe}),groupEnd:g({},e,{value:ie})})}E<0&&m(\"disabledDepth fell below zero. This is a bug in React. Please file an issue.\")}}var W=R.ReactCurrentDispatcher,V;function C(e,t,r){{if(V===void 0)try{throw Error()}catch(l){var a=l.stack.trim().match(/\\n( *(at )?)/);V=a&&a[1]||\"\"}return`\n`+V+e}}var Y=!1,S;{var $e=typeof WeakMap==\"function\"?WeakMap:Map;S=new $e}function le(e,t){if(!e||Y)return\"\";{var r=S.get(e);if(r!==void 0)return r}var a;Y=!0;var l=Error.prepareStackTrace;Error.prepareStackTrace=void 0;var d;d=W.current,W.current=null,Ve();try{if(t){var i=function(){throw Error()};if(Object.defineProperty(i.prototype,\"props\",{set:function(){throw Error()}}),typeof Reflect==\"object\"&&Reflect.construct){try{Reflect.construct(i,[])}catch(_){a=_}Reflect.construct(e,[],i)}else{try{i.call()}catch(_){a=_}e.call(i.prototype)}}else{try{throw Error()}catch(_){a=_}e()}}catch(_){if(_&&a&&typeof _.stack==\"string\"){for(var o=_.stack.split(`\n`),h=a.stack.split(`\n`),c=o.length-1,f=h.length-1;c>=1&&f>=0&&o[c]!==h[f];)f--;for(;c>=1&&f>=0;c--,f--)if(o[c]!==h[f]){if(c!==1||f!==1)do if(c--,f--,f<0||o[c]!==h[f]){var b=`\n`+o[c].replace(\" at new \",\" at \");return e.displayName&&b.includes(\"<anonymous>\")&&(b=b.replace(\"<anonymous>\",e.displayName)),typeof e==\"function\"&&S.set(e,b),b}while(c>=1&&f>=0);break}}}finally{Y=!1,W.current=d,Ye(),Error.prepareStackTrace=l}var k=e?e.displayName||e.name:\"\",ye=k?C(k):\"\";return typeof e==\"function\"&&S.set(e,ye),ye}function Me(e,t,r){return le(e,!1)}function Le(e){var t=e.prototype;return!!(t&&t.isReactComponent)}function P(e,t,r){if(e==null)return\"\";if(typeof e==\"function\")return le(e,Le(e));if(typeof e==\"string\")return C(e);switch(e){case A:return C(\"Suspense\");case F:return C(\"SuspenseList\")}if(typeof e==\"object\")switch(e.$$typeof){case T:return Me(e.render);case O:return P(e.type,t,r);case I:{var a=e,l=a._payload,d=a._init;try{return P(d(l),t,r)}catch{}}}return\"\"}var j=Object.prototype.hasOwnProperty,ue={},de=R.ReactDebugCurrentFrame;function U(e){if(e){var t=e._owner,r=P(e.type,e._source,t?t.type:null);de.setExtraStackFrame(r)}else de.setExtraStackFrame(null)}function Be(e,t,r,a,l){{var d=Function.call.bind(j);for(var i in e)if(d(e,i)){var o=void 0;try{if(typeof e[i]!=\"function\"){var h=Error((a||\"React class\")+\": \"+r+\" type `\"+i+\"` is invalid; it must be a function, usually from the `prop-types` package, but received `\"+typeof e[i]+\"`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.\");throw h.name=\"Invariant Violation\",h}o=e[i](t,i,a,r,null,\"SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED\")}catch(c){o=c}o&&!(o instanceof Error)&&(U(l),m(\"%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).\",a||\"React class\",r,i,typeof o),U(null)),o instanceof Error&&!(o.message in ue)&&(ue[o.message]=!0,U(l),m(\"Failed %s type: %s\",r,o.message),U(null))}}}var qe=Array.isArray;function $(e){return qe(e)}function ze(e){{var t=typeof Symbol==\"function\"&&Symbol.toStringTag,r=t&&e[Symbol.toStringTag]||e.constructor.name||\"Object\";return r}}function Ge(e){try{return ce(e),!1}catch{return!0}}function ce(e){return\"\"+e}function fe(e){if(Ge(e))return m(\"The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.\",ze(e)),ce(e)}var N=R.ReactCurrentOwner,Ke={key:!0,ref:!0,__self:!0,__source:!0},pe,me,M;M={};function Xe(e){if(j.call(e,\"ref\")){var t=Object.getOwnPropertyDescriptor(e,\"ref\").get;if(t&&t.isReactWarning)return!1}return e.ref!==void 0}function He(e){if(j.call(e,\"key\")){var t=Object.getOwnPropertyDescriptor(e,\"key\").get;if(t&&t.isReactWarning)return!1}return e.key!==void 0}function Je(e,t){if(typeof e.ref==\"string\"&&N.current&&t&&N.current.stateNode!==t){var r=w(N.current.type);M[r]||(m('Component \"%s\" contains the string ref \"%s\". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref',w(N.current.type),e.ref),M[r]=!0)}}function Ze(e,t){{var r=function(){pe||(pe=!0,m(\"%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)\",t))};r.isReactWarning=!0,Object.defineProperty(e,\"key\",{get:r,configurable:!0})}}function Qe(e,t){{var r=function(){me||(me=!0,m(\"%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)\",t))};r.isReactWarning=!0,Object.defineProperty(e,\"ref\",{get:r,configurable:!0})}}var et=function(e,t,r,a,l,d,i){var o={$$typeof:n,type:e,key:t,ref:r,props:i,_owner:d};return o._store={},Object.defineProperty(o._store,\"validated\",{configurable:!1,enumerable:!1,writable:!0,value:!1}),Object.defineProperty(o,\"_self\",{configurable:!1,enumerable:!1,writable:!1,value:a}),Object.defineProperty(o,\"_source\",{configurable:!1,enumerable:!1,writable:!1,value:l}),Object.freeze&&(Object.freeze(o.props),Object.freeze(o)),o};function tt(e,t,r,a,l){{var d,i={},o=null,h=null;r!==void 0&&(fe(r),o=\"\"+r),He(t)&&(fe(t.key),o=\"\"+t.key),Xe(t)&&(h=t.ref,Je(t,l));for(d in t)j.call(t,d)&&!Ke.hasOwnProperty(d)&&(i[d]=t[d]);if(e&&e.defaultProps){var c=e.defaultProps;for(d in c)i[d]===void 0&&(i[d]=c[d])}if(o||h){var f=typeof e==\"function\"?e.displayName||e.name||\"Unknown\":e;o&&Ze(i,f),h&&Qe(i,f)}return et(e,o,h,l,a,N.current,i)}}var L=R.ReactCurrentOwner,he=R.ReactDebugCurrentFrame;function x(e){if(e){var t=e._owner,r=P(e.type,e._source,t?t.type:null);he.setExtraStackFrame(r)}else he.setExtraStackFrame(null)}var B;B=!1;function q(e){return typeof e==\"object\"&&e!==null&&e.$$typeof===n}function be(){{if(L.current){var e=w(L.current.type);if(e)return`\n\nCheck the render method of \\``+e+\"`.\"}return\"\"}}function rt(e){{if(e!==void 0){var t=e.fileName.replace(/^.*[\\\\\\/]/,\"\"),r=e.lineNumber;return`\n\nCheck your code at `+t+\":\"+r+\".\"}return\"\"}}var we={};function nt(e){{var t=be();if(!t){var r=typeof e==\"string\"?e:e.displayName||e.name;r&&(t=`\n\nCheck the top-level render call using <`+r+\">.\")}return t}}function _e(e,t){{if(!e._store||e._store.validated||e.key!=null)return;e._store.validated=!0;var r=nt(t);if(we[r])return;we[r]=!0;var a=\"\";e&&e._owner&&e._owner!==L.current&&(a=\" It was passed a child from \"+w(e._owner.type)+\".\"),x(e),m('Each child in a list should have a unique \"key\" prop.%s%s See https://reactjs.org/link/warning-keys for more information.',r,a),x(null)}}function ve(e,t){{if(typeof e!=\"object\")return;if($(e))for(var r=0;r<e.length;r++){var a=e[r];q(a)&&_e(a,t)}else if(q(e))e._store&&(e._store.validated=!0);else if(e){var l=Se(e);if(typeof l==\"function\"&&l!==e.entries)for(var d=l.call(e),i;!(i=d.next()).done;)q(i.value)&&_e(i.value,t)}}}function at(e){{var t=e.type;if(t==null||typeof t==\"string\")return;var r;if(typeof t==\"function\")r=t.propTypes;else if(typeof t==\"object\"&&(t.$$typeof===T||t.$$typeof===O))r=t.propTypes;else return;if(r){var a=w(t);Be(r,e.props,\"prop\",a,e)}else if(t.PropTypes!==void 0&&!B){B=!0;var l=w(t);m(\"Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?\",l||\"Unknown\")}typeof t.getDefaultProps==\"function\"&&!t.getDefaultProps.isReactClassApproved&&m(\"getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.\")}}function ot(e){{for(var t=Object.keys(e.props),r=0;r<t.length;r++){var a=t[r];if(a!==\"children\"&&a!==\"key\"){x(e),m(\"Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.\",a),x(null);break}}e.ref!==null&&(x(e),m(\"Invalid attribute `ref` supplied to `React.Fragment`.\"),x(null))}}function it(e,t,r,a,l,d){{var i=Ie(e);if(!i){var o=\"\";(e===void 0||typeof e==\"object\"&&e!==null&&Object.keys(e).length===0)&&(o+=\" You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.\");var h=rt(l);h?o+=h:o+=be();var c;e===null?c=\"null\":$(e)?c=\"array\":e!==void 0&&e.$$typeof===n?(c=\"<\"+(w(e.type)||\"Unknown\")+\" />\",o=\" Did you accidentally export a JSX literal instead of a component?\"):c=typeof e,m(\"React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s\",c,o)}var f=tt(e,t,r,l,d);if(f==null)return f;if(i){var b=t.children;if(b!==void 0)if(a)if($(b)){for(var k=0;k<b.length;k++)ve(b[k],e);Object.freeze&&Object.freeze(b)}else m(\"React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.\");else ve(b,e)}return e===y?ot(f):at(f),f}}var st=it;G.Fragment=y,G.jsxDEV=st})()});var Ne=z((kt,Ee)=>{\"use strict\";Ee.exports=ke()});var yt={};pt(yt,{default:()=>_t,frontmatter:()=>bt});var s=mt(Ne()),bt={date:new Date(16505856e5),title:\"Code, people and fallibility\",lede:\"Tech failures, like Facebook's DNS mishap and Atlassian's outage, highlight a key insight \\u2014 it's not just about the code, but the humans behind it. These incidents underscore the importance of communication and collaboration in technology, reminding us that at the core of every technical glitch is a human error waiting to be understood and addressed.\",category:\"writing\",tags:[\"teamwork\",\"engineering\"]};function Te(u){let n=Object.assign({p:\"p\",a:\"a\",blockquote:\"blockquote\",em:\"em\",h2:\"h2\"},u.components),{Note:p}=n;return p||vt(\"Note\",!0,\"41:1-44:8\"),(0,s.jsxDEV)(s.Fragment,{children:[(0,s.jsxDEV)(n.p,{children:[\"Last year, Facebook accidentally overwrote its own DNS records leading to a \\u201Ccascade of errors\\u201D and worldwide outage for all its platforms. The outage was likely caused by Facebook network engineers accidentally locking themselves out of the larger configuration system during an update. \",(0,s.jsxDEV)(n.a,{href:\"https://news.ycombinator.com/item?id=28749244\",children:\"This quote\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:11,columnNumber:289},this),\", purportedly from a member of the recovery team, is particularly insightful:\"]},void 0,!0,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:11,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.blockquote,{children:[`\n`,(0,s.jsxDEV)(n.p,{children:\"\\u2026the people with physical access is separate from the people with knowledge of how to actually authenticate to the systems and people who know what to actually do, so there is now a logistical challenge with getting all that knowledge unified.\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:13,columnNumber:3},this),`\n`]},void 0,!0,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:13,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.p,{children:[\"In the last fortnight, \",(0,s.jsxDEV)(n.a,{href:\"https://newsletter.pragmaticengineer.com/p/scoop-atlassian?s=r\",children:\"Atlassian had the longest outage they\\u2019ve ever had\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:15,columnNumber:24},this),\". What started out as \\u201Cdowntime\\u201D on April 4, turned into 10 days of complete shutdown. And they took until \",(0,s.jsxDEV)(n.a,{href:\"https://twitter.com/SriViswan/status/1514072080181956608?s=20&t=VcElOD_KVnRCdgNOLeB5Ig\",children:\"day 9 to formally communicate with their customers\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:15,columnNumber:246},this),\". We can only guess at the chaos behind the scenes.\"]},void 0,!0,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:15,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.p,{children:[\"Of course, we can say that\\u2019s Big Tech, they should have the resources and bright people to figure this out and stop this happening. Shame on them. Or we can say that that's Big Tech, that's different to the scale of what we do. It would never happen to us. And yet if we pause for a minute and consider all the mundane moments that occurred in the prior days, months and perhaps years that lead to these incidents, we may gain a unitary insight into \",(0,s.jsxDEV)(n.em,{children:\"how organisations of people work with code\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:17,columnNumber:451},this),\".\"]},void 0,!0,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:17,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.h2,{children:\"The world of code\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:19,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.p,{children:\"A piece of code takes an input to produce an output and if you add the wrong input, that code itself is none the wiser. But just above that piece of code, we have a world in which the human organisation and the people involved provide the wherewithal for the code to be written into a codebase: what it\\u2019s for, how it works, its dependent inputs, the chain of its related outputs and so on.\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:21,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.p,{children:\"Now picture in your mind how that codebase relates to the larger system at play: the intent of the people in the organisation, their multiple, mixed and shared drivers for what they do, how they co-operate and the end customers they\\u2019re beholden to.\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:23,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.p,{children:\"If we consider the world of code in which the Atlassian Incident took place, we might ask ourselves a few simple questions: did they safeguard the kind of input for this piece of code? Were the people doing the work tired and simply made a mistake? Were there pre-existing issues that had been acknowledged among the group? Was this communicated and with which group?\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:25,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.p,{children:[\"Once we consider questions like these, it is easy to see that \",(0,s.jsxDEV)(n.em,{children:\"the code itself is but one small piece in a larger puzzle that is mostly about human agency and co-operation\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:27,columnNumber:63},this),\".\"]},void 0,!0,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:27,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.p,{children:[\"Because it is so malleable, we can do all kinds of new things with code, some of which have been \",(0,s.jsxDEV)(n.a,{href:\"http://web.archive.org/web/19980610100009/www.redherring.com/mag/issue55/economics.html\",children:\"surprisingly valuable\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:29,columnNumber:98},this),\" for the communications and business spheres in the past three decades. And precisely because it is so malleable, code always contains the possibility of surprising us towards the other side of the reward spectrum: \",(0,s.jsxDEV)(n.em,{children:\"risk\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:29,columnNumber:425},this),\".\"]},void 0,!0,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:29,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.h2,{children:\"The human code\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:31,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.p,{children:\"Beyond a certain point\\u2014as Atlassian and Facebook will attest to\\u2014what is complicated can turn into unbounded complexity. It is easy to gawk at the runaway train of incidents resulting from such complexity until you too wish to make use of what code might do for you. Then you will engage in your own \\u201Cworld of code\\u201D: a place where code does exactly what you and your organisation tell it to, in order to reap the returns across the risk/reward spectrum.\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:33,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.p,{children:[\"How then do we mitigate risk in the world of code? Our best advice is to understand that \",(0,s.jsxDEV)(n.em,{children:\"behind the code is always a fallible human who understands how the code works\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:35,columnNumber:90},this),\". And that human has to co-ordinate and communicate with other fallible humans around their ideas of what that code should do.\"]},void 0,!0,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:35,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.p,{children:\"In such scenarios, we should be humble not just about our ability to programme complicated systems, but also about our ability to mutually understand, trust and co-ordinate with each-other.\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:37,columnNumber:1},this),`\n`,(0,s.jsxDEV)(n.p,{children:\"A unitary insight into working with code is that it\\u2019s not about code, it\\u2019s about people.\"},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:39,columnNumber:1},this),`\n`,(0,s.jsxDEV)(p,{children:(0,s.jsxDEV)(n.p,{children:[\"This article was originally published on the \",(0,s.jsxDEV)(n.a,{href:\"https://grade.substack.com/p/code-people-and-fallibility\",children:`Grade\nSubstack`},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:42,columnNumber:48},this),\".\"]},void 0,!0,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:42,columnNumber:3},this)},void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:41,columnNumber:1},this)]},void 0,!0,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\",lineNumber:1,columnNumber:1},this)}function wt(u={}){let{wrapper:n}=u.components||{};return n?(0,s.jsxDEV)(n,Object.assign({},u,{children:(0,s.jsxDEV)(Te,u,void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\"},this)}),void 0,!1,{fileName:\"/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx\"},this):Te(u)}var _t=wt;function vt(u,n,p){throw new Error(\"Expected \"+(n?\"component\":\"object\")+\" `\"+u+\"` to be defined: you likely forgot to import, pass, or provide it.\"+(p?\"\\nIt\\u2019s referenced in your code at `\"+p+\"` in `/Users/callumflack/Repos/t-tw/apps/web/posts/_mdx_bundler_entry_point-2e2df008-54ab-46d1-b1b0-7e6e998ddf8e.mdx`\":\"\"))}return ht(yt);})();\n/*! Bundled license information:\n\nreact/cjs/react-jsx-dev-runtime.development.js:\n  (**\n   * @license React\n   * react-jsx-dev-runtime.development.js\n   *\n   * Copyright (c) Facebook, Inc. and its affiliates.\n   *\n   * This source code is licensed under the MIT license found in the\n   * LICENSE file in the root directory of this source tree.\n   *)\n*/\n;return Component;"
  },
  "_id": "code-people-and-fallibility.mdx",
  "_raw": {
    "sourceFilePath": "code-people-and-fallibility.mdx",
    "sourceFileName": "code-people-and-fallibility.mdx",
    "sourceFileDir": ".",
    "contentType": "mdx",
    "flattenedPath": "code-people-and-fallibility"
  },
  "type": "Post",
  "slug": "code-people-and-fallibility",
  "readingTime": {
    "text": "4 min read",
    "minutes": 3.545,
    "time": 212700,
    "words": 709
  }
}